src/Trie.mo

```diff
@@ -1517,12 +1545,34 @@ module {
   /// ```motoko include=initialize
   /// trie := Trie.put(trie, key "hello", Text.equal, 42).0;
   /// trie := Trie.put(trie, key "bye", Text.equal, 32).0;
-  /// // remove the value associated with "hello"
+  /// // remove the entry associated with "hello"
   /// trie := Trie.remove(trie, key "hello", Text.equal).0;
   /// assert (Trie.get(trie, key "hello", Text.equal) == null);
   /// ```
-  public func remove<K, V>(t : Trie<K, V>, k : Key<K>, k_eq : (K, K) -> Bool) : (Trie<K, V>, ?V) {
-    replace(t, k, k_eq, null)
+  public func remove<K, V>(trie : Trie<K, V>, key : Key<K>, equal : (K, K) -> Bool) : (Trie<K, V>, ?V) {
+    func rec(trie : Trie<K, V>, bitPosition : Nat) : (Trie<K, V>, ?V) {
+      switch trie {
+        case (#empty) { (#empty, null) };
```

Comment:
This looks plausible to me, but in
https://github.com/dfinity/motoko-base/issues/427#issuecomment-1319671589
Joachim warns against a broken path compression optimization in filter or map. Does that warning apply here?

Reply:
Indeed. The path cannot be compressed for branch nodes because the bit positions matter (as Joachim correctly mentioned). This is why I only collapse the leaf nodes (with an empty sibling or another sibling leaf). Or do I miss something?

Reply:
I added collapsing two empty leaves (since `leaf()` function also returns an empty node for an empty leaf).
After deleting all elements in a 1000_000 trie, the trie is again empty (no branches contained any longer).